### PR TITLE
C++ Demo - Facial Expression Recognition

### PR DESCRIPTION
**Pull request by DaniAffCH**
This Pull Request introduces a C++ demo for utilizing and visualizing the Facial Expression Recognition (FER) model requested in issue #135 . The changes include:

- `demo.cpp`: This file contains the demo implementation, featuring a class `YuNet` for face detection (adapted from `face_detection_yunet/demo.cpp` to create a stand-alone demo) and a class `FER` for preprocessing and inference with the Facial Expression Recognition model.

- `CMakeLists.txt`: The related CMakeLists file necessary for building the program.

- `README.md`: The README file has been updated to include information about this new demo.

The demo offers two modes of inference: real-time using the camera or processing a static image.

The coding style in `demo.cpp` aligns with the conventions used in other demos and follows the structure of the corresponding `demo.py`. Notably, I opted to utilize the `estimateAffine2D` function from the OpenCV library instead of reimplementing a similar mechanism, as was done in the Python version.

## Testing 

The program has been tested both with a single person and with multiple people at the same time.
Here are some screenshots:

#### Single Person from the cam
Screenshot from 2024-02-25 22-56-10

#### Multi-person from static image
result

#### Build & Run
cli run

### CODE REVIEW

File: models/facial_expression_recognition/demo.cpp [hunk 1]
```diff
@@ -0,0 +1,300 @@
...
+    Scalar imageStd = Scalar(0.5,0.5,0.5);
+
+    const String inputNames = "data";
+    const String outputNames = "label";
+
+    int backend_id;
+    int target_id;
+    
+public:
+    FER(const string& modelPath, int backend_id, int target_id) : modelPath(modelPath), backend_id(backend_id), target_id(target_id)
```
Comment (fengyuentau) [inline models/facial_expression_recognition/demo.cpp:48]: Could you give `backend_id` and `target_id` default values like `YuNet` does?

File: models/facial_expression_recognition/demo.cpp [hunk 2]
```diff
@@ -0,0 +1,300 @@
...
+"{ input i          |                                                                  | Path to input image or video file. Skip this argument to capture frames from a camera.}"
+"{ backend_target t | 0                                                                | Choose one of the backend-target pair to run this demo:\n"
+                                                                                        "0: (default) OpenCV implementation + CPU,\n"
+                                                                                        "1: CUDA + GPU (CUDA),\n"
+                                                                                        "2: CUDA + GPU (CUDA FP16),\n"
+                                                                                        "3: TIM-VX + NPU,\n"
+                                                                                        "4: CANN + NPU}"
+"{ save s           | false                                                             | Specify to save results.}"
+"{ vis v            | false                                                             | Specify to open a window for result visualization.}"
```
Comment (fengyuentau) [inline models/facial_expression_recognition/demo.cpp:205]: I propose to set `vis` defaults to true for a straight impression in a first run.

File: models/facial_expression_recognition/demo.cpp [hunk 3]
```diff
@@ -0,0 +1,300 @@
...
+            cout << "Results are saved to result.jpg" << endl;
+
+            cv::imwrite("result.jpg", res_frame);
+        }
+    }
+    
+
+    return 0;
+
+}
```
Comment (fengyuentau) [inline models/facial_expression_recognition/demo.cpp:300]: Put an empty line before file ends.

**Review by fengyuentau — COMMENTED** [review]
Nice and clean 👍 We can proceed to merge after the following minor comments are solved.

**Comment by DaniAffCH** [comment]
Hi @fengyuentau thank you for your review. I just fixed the minor issues you pointed out.

File: models/facial_expression_recognition/demo.cpp [hunk 4]
```diff
@@ -0,0 +1,304 @@
...
+            String exp = expressionRecognitionModel.infer(frame, face);
+            expressions.push_back(exp);
+
+            int x1 = static_cast<int>(faces.at<float>(i, 0));
+            int y1 = static_cast<int>(faces.at<float>(i, 1));
+            int w = static_cast<int>(faces.at<float>(i, 2));
+            int h = static_cast<int>(faces.at<float>(i, 3));
+            float conf = faces.at<float>(i, 14);
+
+            std::cout << cv::format("%d: x1=%d, y1=%d, w=%d, h=%d, conf=%.4f expression=%s\n", i, x1, y1, w, h, conf, exp.c_str());
```
Comment (fengyuentau) [inline models/facial_expression_recognition/demo.cpp:280]: I believe this is a C++20 feature. We use C++11 as same as OpenCV 4.x does. Please also add

```cmake
set(CMAKE_CXX_STANDARD 11)
```

in the CMake script.

Comment (DaniAffCH) [reply models/facial_expression_recognition/demo.cpp:280]: Do you mean the `format`?

**Comment by fengyuentau** [comment]
Oops, sorry. I thought it was std::format. Please ignore this comment.
________________________________
发件人: Daniele Affinita ***@***.***>
发送时间: Monday, February 26, 2024 9:11:08 PM
收件人: opencv/opencv_zoo ***@***.***>
抄送: Yuantao Feng ***@***.***>; Mention ***@***.***>
主题: Re: [opencv/opencv_zoo] C++ Demo - Facial Expression Recognition (PR #233)


@DaniAffCH commented on this pull request.

________________________________

In models/facial_expression_recognition/demo.cpp<https://github.com/opencv/opencv_zoo/pull/233#discussion_r1502589679>:

> +        Mat faces = faceDetectionModel.infer(frame);
+        vector<String> expressions;
+
+        for (int i = 0; i < faces.rows; ++i)
+        {
+            Mat face = faces.row(i);
+            String exp = expressionRecognitionModel.infer(frame, face);
+            expressions.push_back(exp);
+
+            int x1 = static_cast<int>(faces.at<float>(i, 0));
+            int y1 = static_cast<int>(faces.at<float>(i, 1));
+            int w = static_cast<int>(faces.at<float>(i, 2));
+            int h = static_cast<int>(faces.at<float>(i, 3));
+            float conf = faces.at<float>(i, 14);
+
+            std::cout << cv::format("%d: x1=%d, y1=%d, w=%d, h=%d, conf=%.4f expression=%s\n", i, x1, y1, w, h, conf, exp.c_str());


Do you mean the format?

―

**Comment by ShiqiYu** [comment]
@DaniAffCH Thank you for the contribution.

**Pull request merged**